src/streamlink/plugins/twitch.py

```diff
@@ -128,6 +135,79 @@
 )
 
 
+Segment = namedtuple("Segment", "uri duration title key discontinuity scte35 byterange date map")
+
+
+class TwitchM3U8Parser(M3U8Parser):
+    def __init__(self, base_uri=None, worker=None, **kwargs):
+        M3U8Parser.__init__(self, base_uri=base_uri, **kwargs)
+        self.worker = worker
+
+    def parse_tag_ext_x_scte35_out(self, value):
+        self.worker.scte35 = True
+
+    def parse_tag_ext_x_scte35_in(self, value):
+        self.worker.scte35 = False
+
+    def get_segment(self, uri):
+        byterange = self.state.pop("byterange", None)
+        extinf = self.state.pop("extinf", (0, None))
+        date = self.state.pop("date", None)
+        map_ = self.state.get("map")
+        key = self.state.get("key")
+        discontinuity = self.state.pop("discontinuity", False)
+
+        return Segment(
+            uri,
+            extinf[0],
+            extinf[1],
+            key,
+            discontinuity,
+            self.worker.scte35,
+            byterange,
+            date,
+            map_
+        )
+
+
+class TwitchHLSStreamWorker(HLSStreamWorker):
+    scte35 = False
+
+    def _reload_playlist(self, text, url):
+        return load_hls_playlist(text, url, parser=TwitchM3U8Parser, worker=self)
+
+
+class TwitchHLSStreamWriter(HLSStreamWriter):
+    def __init__(self, *args, **kwargs):
+        HLSStreamWriter.__init__(self, *args, **kwargs)
+        self.disable_ads = self.reader.stream.options.get("disable-ads") or False
+        if self.disable_ads:
+            log.debug("Will skip ad segments")
+
+    def fetch(self, sequence, retries=None):
+        if self.closed or not retries:
+            return
+        if self.disable_ads and sequence.segment.scte35:
```

Comment:
Btw, regarding the `HLSStreamWriter`'s `fetch` method, why do segments get filtered out via the `--hls-segment-ignore-names` parameter after building the request parameters?
https://github.com/streamlink/streamlink/blob/42105d82b93bd2250c42af482fa3ba780c2f75ab/src/streamlink/stream/hls.py#L100-L118
I wanted to refactor this method as well, but wasn't sure because of this
https://github.com/streamlink/streamlink/blob/42105d82b93bd2250c42af482fa3ba780c2f75ab/src/streamlink/stream/hls.py#L82-L84

Reply:
> also `kwargs` is not needed

Where? The `HLSStreamWriter` passes both through to `SegmentedStreamWriter`.

> it could simply done with `options.get` like in

Plugin specific options are not set on the session AFAIK (couldn't make it work). Hence this commit
https://github.com/streamlink/streamlink/pull/2372/commits/714ce9f60cdcf3e17e3fe6fb674e5593f5f1f499

Reply:
Nevermind, don't know what I did there, it's properly using the session options now and 714ce9f has been removed from the PR.